modules/gapi/src/compiler/passes/transformations.cpp

```diff
@@ -0,0 +1,140 @@
+void checkCompatibility(const Graph& pattern,
+                        const Graph& substitute,
+                        const Protocol& patternP,
+                        const Protocol& substituteP)
+{
+    const auto& patternDataInputs = patternP.in_nhs;
+    const auto& patternDataOutputs = patternP.out_nhs;
+
+    const auto& substituteDataInputs = substituteP.in_nhs;
+    const auto& substituteDataOutputs = substituteP.out_nhs;
+
+    // number of data nodes must be the same
+    GAPI_Assert(patternDataInputs.size() == substituteDataInputs.size());
+    GAPI_Assert(patternDataOutputs.size() == substituteDataOutputs.size());
+
+    // for each pattern input node, verify a corresponding substitute input node
+    checkDataNodes(pattern, substitute, patternDataInputs, substituteDataInputs);
+
+    // for each pattern output node, verify a corresponding substitute output node
+    checkDataNodes(pattern, substitute, patternDataOutputs, substituteDataOutputs);
+}
+
+// Tries to substitute __single__ pattern with substitute in the given graph
+bool tryToSubstitute(ade::Graph& main,
+                     const std::unique_ptr<ade::Graph>& patternG,
+                     const cv::GComputation& substitute)
+{
+    GModel::Graph gm(main);
+
+    // 1. find a pattern in main graph
+    auto match1 = findMatches(*patternG, gm);
+    if (!match1.ok()) {
+        return false;
+    }
+
+    // 2. build substitute graph inside the main graph
+    cv::gimpl::GModelBuilder builder(main);
+    const auto& proto_slots = builder.put(substitute.priv().m_ins, substitute.priv().m_outs);
+    Protocol substituteP;
+    std::tie(substituteP.inputs, substituteP.outputs, substituteP.in_nhs, substituteP.out_nhs) =
+        proto_slots;
+
+    const Protocol& patternP = GModel::Graph(*patternG).metadata().get<Protocol>();
+
+    // 3. check that pattern and substitute are compatible
+    // FIXME: in theory, we should always have compatible pattern/substitute. if not, we're in
+    //        half-completed state where some transformations are already applied - what can we do
+    //        to handle the situation better?  -- use transactional API as in fuse_islands pass?
+    checkCompatibility(*patternG, gm, patternP, substituteP);
+
+    // 4. make substitution
+    performSubstitution(gm, patternP, substituteP, match1);
+
+    return true;
+}
+}  // anonymous namespace
+
+void applyTransformations(ade::passes::PassContext& ctx,
+                          const gapi::GKernelPackage& pkg,
+                          const std::vector<std::unique_ptr<ade::Graph>>& patterns)
+{
+    const auto& transforms = pkg.get_transformations();
+    const auto size = transforms.size();
+    if (0u == size) return;
+    // Note: patterns are already generated at this point
+    GAPI_Assert(patterns.size() == transforms.size());
+
+    // transform as long as it is possible
+    // Note: check_transformations pass must handle endless loops and such
```

Comment:
Just question: what does "and such" mean?
And check_transformations is not pass anymore..

Reply:
True, will delete this part